### PR TITLE
[katran]: rework ipip healthchecks

### PR DESCRIPTION
**Pull request by tehnerd**
Big diff. Major changes:
- renaming XdpTester to BpfTester + adding features to unittest
BPF_PROG_TYPE_SCHED_CLS (TC bpf).
- adding into katran tester fixtures for healthchecking bpf program
(right now input and output are exactly the same. for now it is
being used either to make sure that prog is loaded and running +
somark in bpf program could be manually changed to some static value
instead of somark=skb->mark (this is till
https://www.spinics.net/lists/netdev/msg620553.html would be available
in vanila kernel)
- adding new healthchecking bpf program (this one forced to reshuffle
some existing code, so existing helpers would be bpf program agnostic
(e.g. before they were including balancer_maps etc)). the new
healthchecking program allow to directly create encapsulation
of the healthcheck packet instead of relying on ip4 and ipv6 tunnel
interfaces w/ external flag set.

Tested-By: katran_tester. for healthchecking somark were manually set to
2 and 3 (to force ipip and ip(4)ip6 healthchecks generation).

ipip:
```
23:12:37.392524 01:00:00:00:00:00 > 02:00:00:00:00:00, ethertype IPv4 (0x0800), length 57: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto UDP (17), length 43)
    192.168.1.1.31337 > 10.200.1.1.80: [udp sum ok] UDP, length 15
23:12:37.392535 00:ff:de:ad:be:af > 00:00:de:ad:be:af, ethertype IPv4 (0x0800), length 77: (tos 0x0, ttl 64, id 0, offset 0, flags [none], proto IPIP (4), length 63)
    10.0.13.37 > 10.0.0.2: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto UDP (17), length 43)
    192.168.1.1.31337 > 10.200.1.1.80: [udp sum ok] UDP, length 15
23:12:37.392655 01:00:00:00:00:00 > 02:00:00:00:00:00, ethertype IPv4 (0x0800), length 69: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto TCP (6), length 55)
    192.168.1.1.31337 > 10.200.1.1.80: Flags [.], cksum 0x27e4 (correct), seq 0:15, ack 1, win 8192, length 15: HTTP
23:12:37.392664 00:ff:de:ad:be:af > 00:00:de:ad:be:af, ethertype IPv4 (0x0800), length 89: (tos 0x0, ttl 64, id 0, offset 0, flags [none], proto IPIP (4), length 75)
    10.0.13.37 > 10.0.0.2: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto TCP (6), length 55)
    192.168.1.1.31337 > 10.200.1.1.80: Flags [.], cksum 0x27e4 (correct), seq 0:15, ack 1, win 8192, length 15: HTTP
```

ip(4)ip6
```
23:17:54.987396 01:00:00:00:00:00 > 02:00:00:00:00:00, ethertype IPv4 (0x0800), length 57: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto UDP (17), length 43)
    192.168.1.1.31337 > 10.200.1.1.80: [udp sum ok] UDP, length 15
23:17:54.987408 00:ff:de:ad:be:af > 00:00:de:ad:be:af, ethertype IPv6 (0x86dd), length 97: (hlim 64, next-header IPIP (4) payload length: 43) fc00:2307::1337 > fc00::1: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto UDP (17), length 43)
    192.168.1.1.31337 > 10.200.1.1.80: [udp sum ok] UDP, length 15
23:17:54.987422 01:00:00:00:00:00 > 02:00:00:00:00:00, ethertype IPv4 (0x0800), length 69: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto TCP (6), length 55)
    192.168.1.1.31337 > 10.200.1.1.80: Flags [.], cksum 0x27e4 (correct), seq 0:15, ack 1, win 8192, length 15: HTTP
23:17:54.987426 00:ff:de:ad:be:af > 00:00:de:ad:be:af, ethertype IPv6 (0x86dd), length 109: (hlim 64, next-header IPIP (4) payload length: 55) fc00:2307::1337 > fc00::1: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto TCP (6), length 55)
    192.168.1.1.31337 > 10.200.1.1.80: Flags [.], cksum 0x27e4 (correct), seq 0:15, ack 1, win 8192, length 15: HTTP
23:17:54.987439 01:00:00:00:00:00 > 02:00:00:00:00:00, ethertype IPv6 (0x86dd), length 89: (hlim 64, next-header TCP (6) payload length: 35) fc00:2::1.31337 > fc00:1::1.80: Flags [.], cksum 0xfd4f (correct), seq 0:15, ack 1, win 8192, length 15: HTTP
23:17:54.987443 00:ff:de:ad:be:af > 00:00:de:ad:be:af, ethertype IPv6 (0x86dd), length 129: (hlim 64, next-header IPv6 (41) payload length: 75) fc00:2307::1337 > fc00::1: (hlim 64, next-header TCP (6) payload length: 35) fc00:2::1.31337 > fc00:1::1.80: Flags [.], cksum 0xfd4f (correct), seq 0:15, ack 1, win 8192, length 15: HTTP
```

### CODE REVIEW

**Comment by tehnerd** [comment]
force pushed w/ addition to ability to change destination mac for healthchecks through changeMac call

File: katran/lib/bpf/healthchecking_kern.c [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#include "healthchecking_helpers.h"
+#include "healthchecking_maps.h"
+
+
+SEC("cls-hc")
+int healthchecker(struct __sk_buff *skb)
+{
+  __u32 stats_key = GENERIC_STATS_INDEX;
+  __u32 somark = skb->mark;
+  __u32 key = 3;
```
Comment (udippant) [inline katran/lib/bpf/healthchecking_kern.c]: nit: explain magic value (comment or descriptive constant)

File: katran/lib/KatranLb.cpp [hunk 2]
```diff
@@ -312,27 +314,64 @@ void KatranLb::setupGueEnvironment() {
...
+  if (res < 0) {
+    throw std::runtime_error("can not update src v4 address for GUE packet");
+  }
+  key = kSrcV6Pos;
+  res = bpfAdapter_.bpfUpdateMap(map_fd, &key, &srcv6);
+  if (res < 0) {
+    throw std::runtime_error("can not update src v6 address for GUE packet");
+  }
+
+  struct hc_mac macs[2];
```
Comment (udippant) [inline katran/lib/KatranLb.cpp]: Suggestion from internal lint - Prefer std::array over C arrays: std::array

File: katran/lib/KatranLb.cpp [hunk 3]
```diff
@@ -312,27 +314,64 @@ void KatranLb::setupGueEnvironment() {
...
+  // populating mac addresses for healthchecking
+  if (config_.localMac.size() != 6) {
+    throw std::invalid_argument("src mac's size is not equal to six byte");
+  }
+  for (int i = 0; i < 6; i++) {
+    macs[kHcSrcMacPos].mac[i] = config_.localMac[i];
+    macs[kHcDstMacPos].mac[i] = config_.defaultMac[i];
+  }
+  for (auto key : {kHcSrcMacPos, kHcDstMacPos}) {
```
Comment (udippant) [inline katran/lib/KatranLb.cpp]: declaration of `key` here shadows the earlier declaration in line # 337.

File: katran/lib/bpf/healthchecking_helpers.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    __u8 proto = IPPROTO_IPV6;
+    key = V6_SRC_INDEX;
+    src = bpf_map_lookup_elem(&hc_pckt_srcs_map, &key);
+    if (!src) {
+      return false;
+    }
+    flags |= BPF_F_ADJ_ROOM_FIXED_GSO | BPF_F_ADJ_ROOM_ENCAP_L3_IPV6;
+    adjust_len = sizeof(struct ipv6hdr);
+    // new header would be inserted after MAC but before old L3 header
+    if(bpf_skb_adjust_room(skb, adjust_len, BPF_ADJ_ROOM_MAC, flags)) {
```
[the remark this answers is not in the document]
Comment (tehnerd) [reply katran/lib/bpf/healthchecking_helpers.h]: This is ipip encap. Gue is separate diff. So no udp in this one

File: katran/lib/KatranLbStructs.h [hunk 5]
```diff
@@ -191,6 +192,7 @@ struct KatranConfig {
   bool memlockUnlimited = true;
   std::string katranSrcV4 = kAddressNotSpecified;
   std::string katranSrcV6 = kAddressNotSpecified;
+  std::vector<uint8_t> localMac;
```
[the remark this answers is not in the document]
Comment (tehnerd) [reply katran/lib/KatranLbStructs.h]: we already using std::vector for remote mac and helpers  are expecting it (vector). i'd prefer not to create new instance and to use the same everywhere: question is - do fb want to break katran's API? (aka to rework remote mac as well)

File: katran/lib/bpf/csum_helpers.h [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef __CSUM_HELPERS_H
+#define __CSUM_HELPERS_H
+
+#include <linux/in.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <stdbool.h>
```
[the remark this answers is not in the document]
Comment (tehnerd) [reply katran/lib/bpf/csum_helpers.h]: i dont remember. this is was mostly copy paste from exisitn .h after shufling code around. i will clean things up

File: katran/lib/bpf/balancer_maps.h [hunk 7]
```diff
@@ -46,7 +46,6 @@ struct bpf_map_def SEC("maps") lru_maps_mapping = {
   .max_entries = MAX_SUPPORTED_CPUS,
   .map_flags = NO_FLAGS,
 };
-BPF_ANNOTATE_KV_PAIR(lru_maps_mapping, __u32, __u32);
```
[the remark this answers is not in the document]
Comment (tehnerd) [reply katran/lib/bpf/balancer_maps.h]: it does not work for map in map (libbpf prints error on loading)

File: katran/lib/bpf/balancer_helpers.h [hunk 8]
```diff
@@ -23,12 +23,14 @@
 #include <linux/in.h>
 #include <linux/ip.h>
 #include <linux/ipv6.h>
+#include <stdbool.h>
```
[the remark this answers is not in the document]
Comment (tehnerd) [reply katran/lib/bpf/balancer_helpers.h]: this is because we are using "bool is_ipv6" in this file

**Comment by tehnerd** [comment]
force pushed new version w/ addressed comments from above:
- c array -> cpp array
- directHealthchecking now in featureDiscovering section
- nits on space / magic numbers.
- reworked include of stdbool.h in bpf code  (we are using it because we have bool type variables)
- fixed shadowed variable


re on PR processing: i'm doing force push because somehow from the top of my head i remember that FB's internal import tool cannot import PR w/ multipe commits. is it still the same? or should i add sep commits if something changed (easier to review on github as you would be able to see what has exactly changed) and then, during the land, it would just squashed to a single commit?

**Pull request closed without merging**